### PR TITLE
Replace allowed_pipelines with allowed_component_graphs

### LINKED ISSUE #2159: Update ``AutoMLSearch`` to accept ``allowed_component_graphs`` instead of ``allowed_pipelines``

### PR DESCRIPTION
**Pull request by ParthivNaresh**
Fixes #2159 and #2166 

[Here](https://alteryx.quip.com/XUzlAfngfhjn) are the docs

Here's an example of the new implementation:
```
from pprint import pp

from evalml.automl import AutoMLSearch
from sklearn.datasets import make_classification

X, y = make_classification(
    n_samples=100, n_features=20, n_informative=2, n_redundant=2, random_state=0
    )

allowed_component_graph = {
        "Logistic Regression Binary Pipeline": [
            "Imputer",
            "One Hot Encoder",
            "Standard Scaler",
            "Logistic Regression Classifier",
        ]
    }
automl = AutoMLSearch(
    X_train=X,
    y_train=y,
    problem_type="binary",
    max_iterations=3,
    allowed_component_graphs=[allowed_component_graph],
    ensembling=True,
)
automl.search()

for pipe in automl.full_rankings['parameters']:
    pp(pipe)

================= OUTPUT ==================
{'Imputer': {'categorical_impute_strategy': 'most_frequent',
             'numeric_impute_strategy': 'mean',
             'categorical_fill_value': None,
             'numeric_fill_value': None},
 'One Hot Encoder': {'top_n': 10,
                     'features_to_encode': None,
                     'categories': None,
                     'drop': 'if_binary',
                     'handle_unknown': 'ignore',
                     'handle_missing': 'error'},
 'Logistic Regression Classifier': {'penalty': 'l2',
                                    'C': 1.0,
                                    'n_jobs': -1,
                                    'multi_class': 'auto',
                                    'solver': 'lbfgs'}}
{'Imputer': {'categorical_impute_strategy': 'most_frequent',
             'numeric_impute_strategy': 'most_frequent',
             'categorical_fill_value': None,
             'numeric_fill_value': None},
 'One Hot Encoder': {'top_n': 10,
                     'features_to_encode': None,
                     'categories': None,
                     'drop': 'if_binary',
                     'handle_unknown': 'ignore',
                     'handle_missing': 'error'},
 'Logistic Regression Classifier': {'penalty': 'l2',
                                    'C': 8.474044870453413,
                                    'n_jobs': -1,
                                    'multi_class': 'auto',
                                    'solver': 'lbfgs'}}
{'Baseline Classifier': {'strategy': 'mode'}}
```

### CODE REVIEW

File: evalml/tests/automl_tests/test_automl.py [hunk 1]
```diff
@@ -1942,12 +1992,24 @@ class Pipeline2(DummyPipeline):
             y_train=y,
             problem_type=problem_type_value,
             max_iterations=3,
-            allowed_pipelines=[Pipeline1({}), Pipeline2({})],
             objective=objective,
             additional_objectives=[],
             optimize_thresholds=False,
             n_jobs=1,
         )
+    automl._automl_algorithm = IterativeAlgorithm(
```
Comment (bchen1116) [inline evalml/tests/automl_tests/test_automl.py:2000]: Did you need to define this since we're patching IterativeAlgo? Why did we not need to do it before?

File: evalml/tests/automl_tests/test_automl.py [hunk 2]
```diff
@@ -2114,6 +2188,23 @@ def fit(self, *args, **kwargs):
         )
 
 
+def test_time_series_regression_with_parameters(ts_data):
```
Comment (bchen1116) [inline evalml/tests/automl_tests/test_automl.py:2191]: Can we not patch this test to make it run faster? We are just making sure that the problem configuration is set properly, right?

File: evalml/tests/automl_tests/test_automl.py [hunk 3]
```diff
@@ -2155,10 +2246,22 @@ def clone(self):
         y_train=y,
         problem_type="binary",
         max_iterations=2,
-        allowed_pipelines=[DummyPipeline({})],
         objective="log loss binary",
         additional_objectives=["f1"],
     )
+    automl._automl_algorithm = IterativeAlgorithm(
```
Comment (bchen1116) [inline evalml/tests/automl_tests/test_automl.py:2252]: Why is this needed?

**Review by bchen1116 — COMMENTED** [review]
Didn't yet finish, but left a few questions that would help me understand! Looking good so far!

File: evalml/automl/automl_algorithm/iterative_algorithm.py [hunk 4]
```diff
@@ -184,11 +181,8 @@ def next_batch(self):
         return next_batch
 
     def _combine_parameters(self, pipeline, proposed_parameters):
-        """Helper function for logic to transform proposed parameters and frozen parameters."""
-        return {
-            **self._transform_parameters(pipeline, proposed_parameters),
-            **self._frozen_pipeline_parameters,
-        }
+        """Helper function for logic to transform proposed parameters."""
```
[the remark this answers is not in the document]
Comment (ParthivNaresh) [reply evalml/automl/automl_algorithm/iterative_algorithm.py:184]: Great point!

File: evalml/tests/automl_tests/test_automl.py (hunk 2 above)
Comment (ParthivNaresh) [reply evalml/tests/automl_tests/test_automl.py:2191]: `search` is actually not even required here, good catch!

File: evalml/tests/automl_tests/test_automl.py (hunk 3 above)
Comment (ParthivNaresh) [reply evalml/tests/automl_tests/test_automl.py:2252]: Because the `DummyPipeline` score is being mocked for the purpose of this test (which is something ComponentGraph doesn't have), I have to specify that pipeline being passed into the `IterativeAlgorithm` since we can't pass it through `AutoMLSearch` anymore. This is the same reason it was added for the other tests as well

File: evalml/automl/automl_search.py [hunk 5]
```diff
@@ -188,10 +188,13 @@ def __init__(
             tolerance (float): Minimum percentage difference to qualify as score improvement for early stopping.
                 Only applicable if patience is not None. Defaults to None.
 
-            allowed_pipelines (list(class)): A list of PipelineBase subclasses indicating the pipelines allowed in the search.
-                The default of None indicates all pipelines for this problem type are allowed. Setting this field will cause
+            allowed_component_graphs (list): A list of dictionaries indicating the component graphs allowed in the search.
+                The format should follow [ {Name_of_graph: [list_of_components]} ].
```
[the remark this answers is not in the document]
Comment (ParthivNaresh) [reply evalml/automl/automl_search.py]: Good point, I think that needs to be integrated!

File: evalml/automl/automl_search.py [hunk 6]
```diff
@@ -188,10 +189,14 @@ def __init__(
...
 
-            allowed_pipelines (list(class)): A list of PipelineBase subclasses indicating the pipelines allowed in the search.
-                The default of None indicates all pipelines for this problem type are allowed. Setting this field will cause
+            allowed_component_graphs (list): A list of dictionaries or ComponentGraphs indicating the component graphs allowed in the search.
+                The format should follow [ {Name_of_graph: [list_of_components],
+                                            random_seed: 42} ].
+                The default of None indicates all pipeline component graphs for this problem type are allowed. Setting this field will cause
                 allowed_model_families to be ignored.
 
+                e.g. allowed_component_graphs = [ {"Name": ["Imputer", "One Hot Encoder", dummy_classifier_estimator_class]} ]
```
Comment (freddyaboulton) [inline evalml/automl/automl_search.py:198]: Nit-pick: I don't think we should use `dummy_classifier_estimator_class` since it's a test fixture.

File: evalml/automl/automl_search.py [hunk 7]
```diff
@@ -378,18 +383,24 @@ def __init__(
...
+                    "Parameter allowed_component_graphs must be either None or a list!"
+                )
+            for graph in allowed_component_graphs:
+                if not isinstance(graph, (dict, ComponentGraph)):
+                    raise ValueError(
+                        "Every component graph passed must be of type dictionary or ComponentGraph!"
+                    )
+            unique_names = set()
+            for graph in allowed_component_graphs:
+                unique_names.add(list(graph.keys())[0])
```
Comment (freddyaboulton) [inline evalml/automl/automl_search.py:398]: I'm wondering if we should change the expected type of `allowed_component_graphs` from a list of dicts to just a dict.

So from 
```python
            allowed_component_graphs=[
                {"Name_0": [dummy_classifier_estimator_class]},
                {"Name_1": [dummy_classifier_estimator_class]},
                {"Name_2": [dummy_classifier_estimator_class]},
            ],
```

to 

```python
            allowed_component_graphs={
                "Name_0": [dummy_classifier_estimator_class],
                "Name_1": [dummy_classifier_estimator_class],
                "Name_2": [dummy_classifier_estimator_class]
},
```

I see there's some duplicated code in this pr to unpack the list, e.g `list(graph.keys())[0]` and `next(iter(component_graph))` making this change may make things clearer/reduce code?

File: evalml/automl/automl_algorithm/iterative_algorithm.py [hunk 8]
```diff
@@ -136,7 +133,7 @@ def next_batch(self):
         if self._batch_number == 0:
             next_batch = [
                 pipeline.new(
-                    parameters=self._combine_parameters(pipeline, {}),
+                    parameters={**self._transform_parameters(pipeline, {})},
```
Comment (freddyaboulton) [inline evalml/automl/automl_algorithm/iterative_algorithm.py:136]: I think in all three lines in this diff we can just do `parameters=self._transform_parameters(pipeline, ...`

File: evalml/tests/automl_tests/test_automl.py [hunk 9]
```diff
@@ -3641,48 +3742,48 @@ def test_automl_respects_pipeline_custom_hyperparameters_with_duplicate_componen
...
     for i, row in automl.full_rankings.iterrows():
         if "Mode Baseline Binary" in row["pipeline_name"]:
             continue
-        if row["pipeline_name"] == "Pipeline Dict":
-            assert row["parameters"]["Imputer"]["numeric_impute_strategy"] in {
-                "most_frequent",
-                "mean",
-            }
+        if row["pipeline_name"] == "Name_linear":
+            assert row["parameters"]["Imputer"]["numeric_impute_strategy"] == "mean"
```
Comment (freddyaboulton) [inline evalml/tests/automl_tests/test_automl.py:3766]: What's causing this diff? Is it the deletion of `_frozen_pipeline_parameters`?

File: evalml/tests/conftest.py [hunk 10]
```diff
@@ -347,6 +347,112 @@ def fit(self, X, y):
     return MockEstimator
 
 
+@pytest.fixture
+def dummy_binary_estimator_class():
```
Comment (freddyaboulton) [inline evalml/tests/conftest.py:351]: Do we need this new fixture? I think we can reuse `dummy_classifier_estimator_class`

**Review by freddyaboulton — APPROVED** [review]
@ParthivNaresh Looks good! I have some minor comments I'd like to discuss before merge though.

File: evalml/tests/automl_tests/test_automl_dask.py [hunk 11]
```diff
@@ -135,10 +161,22 @@ def test_automl_immediate_quit(self):
...
+            tuner_class=SKOptTuner,
+            random_seed=0,
+            n_jobs=-1,
+            number_features=X.shape[1],
+            pipelines_per_batch=5,
+            ensembling=False,
+            text_in_ensembling=False,
+            pipeline_params={},
+            custom_hyperparameters=None,
+        )
```
Comment (chukarsten) [inline evalml/tests/automl_tests/test_automl_dask.py:179]: All these instances of having to set the IterativeAlgorithm are kinda weird.

**Review by chukarsten — APPROVED** [review]
Wow, that was a lot of work.  Good job.  I think we should discuss whether we want to preserve the two different ways of passing in the component graphs that are allowed.  To me, it would seem the way to do it would be to reduce the complexity and only pass in a dict of component graphs (if they need to be named).  The handfulls of instances where IterativeAlgorithm is set in the tests is also weird, but I don't think these are blocking to me.

File: evalml/tests/automl_tests/test_automl.py (hunk 9 above)
Comment (ParthivNaresh) [reply evalml/tests/automl_tests/test_automl.py:3766]: Yes I think so

File: evalml/tests/conftest.py (hunk 10 above)
Comment (ParthivNaresh) [reply evalml/tests/conftest.py:351]: Originally I wanted a fixture that solely supported binary and multiclass each for use cases that are solely dependent on those problem types but the tests pass with `dummy_classifier_estimator_class` anyways so I'll remove those

File: evalml/automl/automl_search.py (hunk 7 above)
Comment (ParthivNaresh) [reply evalml/automl/automl_search.py]: Originally the format was set as such to support `random_seed` but since we're not passing that anymore, this would definitely be a better alternative.
